pkg/awsutils/awsutils.go

```diff
@@ -346,20 +356,106 @@ func (cache *EC2InstanceMetadataCache) initWithEC2Metadata() error {
 	}
 	log.Debugf("Found vpc-ipv4-cidr-block: %s ", cache.vpcIPv4CIDR)
 
-	// retrieve vpc-ipv4-cidr-blocks
+	// initSGIDs retrieves security groups
+	err = cache.refreshSGIDs(mac)
+	if err != nil {
+		return err
+	}
+
+	// initVPCIPv4CIDRs retrieves VPC IPv4 CIDR blocks
+	err = cache.refreshVPCIPv4CIDRs(mac)
+	if err != nil {
+		return err
+	}
+
+	// refresh security groups and VPC CIDR blocks in the background
+	go wait.Forever(func() { cache.refreshSGIDs(mac) }, 30*time.Second)
+	go wait.Forever(func() { cache.refreshVPCIPv4CIDRs(mac) }, 30*time.Second)
+
+	select {
+	case <-ctx.Done():
+		return nil
+	}
+}
+
+// refreshSGIDs retrieves security groups
+func (cache *EC2InstanceMetadataCache) refreshSGIDs(mac string) error {
+	cache.securityGroups.Lock()
+	defer cache.securityGroups.Unlock()
+	metadataSGIDs, err := cache.ec2Metadata.GetMetadata(metadataMACPath + mac + metadataSGs)
+	if err != nil {
+		awsAPIErrInc("GetMetadata", err)
+		log.Errorf("Failed to retrieve security-group-ids data from instance metadata service")
+		return errors.Wrap(err, "get instance metadata: failed to retrieve security-group-ids")
+	}
+
+	sgIDs := strings.Fields(metadataSGIDs)
+
+	if cache.securityGroups.data.Len() == 0 {
+		cache.securityGroups.data = make(sets.String)
+		for _, sgID := range sgIDs {
+			log.Debugf("Found security-group id: %s", sgID)
+			cache.securityGroups.data.Insert(sgID)
+		}
+	}
+
+	if cache.securityGroups.data.Len() != 0 {
+		cache.securityGroups.dataNew = make(sets.String)
+		for _, sgID := range sgIDs {
+			cache.securityGroups.dataNew.Insert(sgID)
+		}
+		//For example: s1 = {a1, a2, a3} s2 = {a1, a2, a4, a5} s1.Difference(s2) = {a3} s2.Difference(s1) = {a4, a5}
+		for delSG, _ := range cache.securityGroups.data.Difference(cache.securityGroups.dataNew) {
+			cache.securityGroups.data.Delete(delSG)
+			log.Debugf("Deleted Security Group: %s", delSG)
+		}
+		for addSG, _ := range cache.securityGroups.dataNew.Difference(cache.securityGroups.data) {
+			cache.securityGroups.data.Insert(addSG)
+			log.Debugf("Added Security Group: %s", addSG)
+		}
+	}
+
+	return nil
+}
```

Comment:
The above code from 403-419 is just setting the `cache.securityGroups.data` value to the set of security groups in the `sgIDs` variable. I don't think that's what you wanted is it?